pkg/aggregator/ckey/key.go

```diff
@@ -27,49 +24,131 @@ import (
 // Note that Agent <= 6.19.0 were using a 128 bits hash, we've switched
 // to 64 bits for better performances (map access) and because 128 bits were overkill
 // in the first place.
-// Note that we've benchmarked against xxhash64 which should be slightly faster,
-// but the Go compiler is not inlining xxhash sum methods whereas it is inlining
-// the murmur3 implementation, providing better performances overall.
-// Note that benchmarks against fnv1a did not provide better performances (no inlining).
+// Note that benchmarks against fnv1a did not provide better performances (no inlining)
+// nor did benchmarks with xxhash (slightly slower).
 type ContextKey uint64
 
-// KeyGenerator generates key
-// Not safe for concurrent usage
-type KeyGenerator struct {
-	buf []byte
-}
+// hashSetSize is the size selected for hashset used to deduplicate the tags
+// while generating the hash. This size has been selected to have space for
+// approximately 500 tags since it's not impacting much the performances,
+// even if the backend is truncating after 100 tags.
+const hashSetSize = 512
+
+// bruteforceSize is the threshold number of tags below which a bruteforce algorithm is
+// faster than a hashset.
+const bruteforceSize = 16
+
+// blank is a marker value to indicate that hashset slot is vacant.
+const blank = -1
 
 // NewKeyGenerator creates a new key generator
 func NewKeyGenerator() *KeyGenerator {
+	empty := make([]int, 0, hashSetSize)
+	for len(empty) < hashSetSize {
+		empty = append(empty, blank)
+	}
 	return &KeyGenerator{
-		buf: make([]byte, 0, 1024),
+		seen:    make([]uint64, hashSetSize),
+		seenIdx: make([]int, hashSetSize),
+		empty:   empty,
 	}
 }
 
+// KeyGenerator generates hash for the given name, hostname and tags.
+// The tags don't have to be sorted and duplicated tags will be ignored while
+// generating the hash.
+// Not safe for concurrent usage.
+type KeyGenerator struct {
+	// reused buffer to not create a uint64 on the stack every key generation
+	intb uint64
+
+	// seen is used as a hashset to deduplicate the tags when there is more than
+	// 16 and less than 512 tags.
+	seen []uint64
+	// seenIdx is the index of the tag stored in the hashset
+	seenIdx []int
+	// empty is an empty hashset with all values set to `blank`, to reset `seenIdx`
+	empty []int
```

Comment:
I asked last time this was changed, but .. I think go uses `memset` properly to clear arrays, which should be substantially faster than copying a known-zero memory region over.

Reply:
Ah, here's the link -- https://github.com/golang/go/commit/f03c9202c43e0abb130669852082117ca50aa9b1

Reply:
We talked about this and it turns out that because this is in a struct, the compiler can't prove that the array is not aliased, so it can't use memclr on it.